### ISSUE TITLE
BeforeOrAfter operator

### ISSUE THREAD

**Issue by aguynamedryan** [outcomesinsights/conceptql#136]
Takes an at_least and within option.

Also takes a checkbox for "allow overlap" to decide if an event must be wholly contained within the before/after range

**Comment by aguynamedryan** [comment]
If I have multiple RHS rows for a patient, how do we want to handle them?

The Before operator works by finding the oldest start_date on the RHS and returns only those LHS events with an end_date earlier than the oldest start_date.  (LHS.end_date < min(RHS.start_date))

The After operator works by finding the most recent end_date on the RHS and returns only those LHS events with a start_date after the most recent end_date. (LHS.start_date > max(RHS.end_date))

So for this operator, do we find all LHS events with an end_date before the earliest RHS start_date along with all LHS events with a start_date after the most recent RHS end_date?  (LHS.end_date < min(RHS.start_date)) OR (LHS.start_date > max(RHS.end_date))

**Comment by markdanese** [comment]
So, it means that "before" is actually "before all", and "after" is really "after all"?  We need to clarify that for users because sometimes people might just want to make sure that X comes before an instance of Y, which is how I was assuming this was operating (not having looked at the documentation).

**Comment by markdanese** [comment]
Should we have two operators for before and after-- "before all" and "before any" and "after all" and "after any"?

I think of "before or after" as "not overlapping".  So, I view it as looking for records on the LHS that don't overlap records on the RHS.  In this case, I would think about it as "not overlapping any" and return all of the LHS records that don't overlap anything from the RHS.  But maybe we should resolve "before" and "after" first, before we try and do anything more complicated.

**Comment by aguynamedryan** [comment]
Whoops.  First off, I misrepresented how Before and After work.  Here's the real deal:

The Before operator works by finding the *most recent* start_date on the RHS and returns only those LHS events with an end_date earlier than the *most recent* start_date. (LHS.end_date < max(RHS.start_date))

The After operator works by finding the *oldest* end_date on the RHS and returns only those LHS events with a start_date earlier than the *oldest* start_date. (LHS.end_date < min(RHS.start_date))

So Before with no other options is "Before Most Recent" and After is "After Oldest".  We already explain the logic of why we do this here: https://github.com/outcomesinsights/conceptql_spec#edge-behaviors

If a user adds a "within" option, then we compare each RHS row against each LHS row because the notion of "before" or "after" is changed to represent a specific range of time.  In this case, Before acts as a "Before Any" and After acts as an "After Any"

Please note that if "at_least" or "occurrences" are specified, we still do the same trick where we find the oldest/newest RHS dates.  If someone wants events that occur, say, at_least 10 days after a set of RHS events, in my mind, any event that occurs 10 days after the earliest event is an event that qualifies.

**Comment by aguynamedryan** [comment]
So what does it look like if we ask if an LHS row is BeforeOrAfter multiple RHS rows?  I'd argue that we find the oldest RHS start_date and the most recent RHS end_date and find all events that fall outside that range.

Or rather:

```
LHS.end_date < min(RHS.start_date) OR LHS.start_date > max(RHS.end_date)
```

Since BeforeOrAfter can take a "within" option, in that situation we'd need to evaluate each LHS against each RHS.

![before 2fafter behavior - page 1](https://user-images.githubusercontent.com/117722/28295179-d401a2e8-6b13-11e7-92a7-14530b26878a.png)

**Comment by markdanese** [comment]
I have to think about this more, but it all looks right to me.  Before and After make sense -- you only need to know the oldest or newest to make the comparison.

Before or After seems to be the equivalent of having an "OR" operator and using it to combine them.    Which makes sense to me.

I think all of these make more sense when one of the sides is reduced to a single event, which you point out in the specs.

**Comment by aguynamedryan** [comment]
Should we bump this?

**Comment by markdanese** [comment]
Yes.  People can use Before and After separately.  It is a little more work but it is doable.

**Comment by jenniferduryea** [comment]
moved it to broom.

**Comment by aguynamedryan** [comment]
I'm awaiting final specifications for this operator.

**Comment by jenniferduryea** [comment]
@markdanese Is this still relevant for Broom or should we move this to Chisel?
